_posts/2021-07-13-JEV.md

```diff
@@ -0,0 +1,822 @@
+use double the memory size. 
+
+The problem is, our naïveté assumes that an array stays where it is, but know this, SIMD or not, a CPU only deals with
+registers.
+
+```java
+final int offset = 0;
+IntVector vector = IntVector.fromArray(species, rgb, offset);
+```
+
+Here we create a _Vector_ of the type and width we determined with the `species` before.
+
+Meaning, if `species == 128 bytes wide`, and we are dealing with `integers`, then `vector` will contain the
+first `4 ints` from the `rgb` array, at the specified `offset`.
+
+###### Do operation(s) on SIMD register(s)
+
+Now that the data is where we want it to be, we can unleash our CPU prowess upon it!!!
+
+As you will no doubt discover when you press ctrl + space on `vector.*`, there's a lot of fun things you can do with the
+vector. These of course are but a sane subset of SIMD, and instead of whining about it, be grateful (and send John Rose
+& Paul Sandoz a gift basket).   
+
+```java
+IntVector result = vector.add(1);
+```
+
+This step is self-explanatory.
+
+###### Write SIMD register(s) back to input
+
+```java
+result.intoArray(rgb, offset);
+```
+
+And finally, we write the updated values back to the array. The way we do it here isn't really the 'proper' way, but it
+works nonetheless. By proper, I mean writing the values back to the input array. It's much better to write to an
+intermediate array (as we might explain later). 
+
+###### Hello hello
+
+This is what our end result looks like:                                                      
+
+```java
+final VectorSpecies<Integer> species = IntVector.SPECIES_PREFERRED;
+int offset = 0;
+for (offset = 0; offset < rgb.length; offset += species.length()) {
+    final IntVector vector = IntVector.fromArray(species, rgb, offset);
+    final IntVector result = vector.add(1);
+    result.intoArray(rgb, offset);
+}
+```
+
+A casual observer might...observe, that we are still looping over our array. This is a necessary evil. Computers are
+magical, but not THAT magical. Maybe someday there will be a single instruction that operates on any length of data.
+Until that day though, loops rule! 
+                            
+And of course this loop might have a tail/remainder, but we will leave it up to your imagination to implement this.
+
+##### Benchmark Vanilla vs Nuts
+
+Is the SIMD version faster or slower? Good question. To be honest, I expected it to be slightly slower because I assumed
+the `Vector` API had some slight overhead compared to the compiler provided `AV`.
+
+Seems I was only half right though, as in it is indeed **a little** slower, but not for the reasons I thought.
+
+So running both versions for 1000 iterations (after warmup) yields the following results:
+
+|vanilla      |simd         |
+|---          |---          | 
+|1249416200ns |1295295700ns |
+
+That's a difference of ~45ms in favor of the vanilla loop.
```

Comment:
40ms on a 1.2s run almost feels like it could be some other overhead. Perhaps you could expand a little bit more on how you ran these tests? (or go completely overboard and could try the JMH framework, if compatible...).